kmod/src/forest.c

```diff
@@ -801,23 +810,162 @@ int scoutfs_forest_oino_alloc(struct super_block *sb, u64 trans_seq)
 	return ret;
 }
 
-void scoutfs_forest_oino_update(struct super_block *sb)
+int scoutfs_forest_oino_update(struct super_block *sb)
 {
 	struct scoutfs_bloom_block *bb;
 	DECLARE_FOREST_INFO(sb, finf);
+	u64 last_seq;
+	int ret;
 	int i;
 
 	if (!finf->oino_bl)
-		return;
+		return 0;
 
 	bb = finf->oino_bl->data;
 
+	ret = scoutfs_client_get_last_seq(sb, &last_seq);
+	WARN_ON_ONCE(ret);
+
+	bb->hdr.seq = cpu_to_le64(last_seq);
+	finf->our_log.oino_bloom_ref.seq = bb->hdr.seq;
+
 	/* Turn the counted BF into a classic BF */
 	for (i = 0; i < SCOUTFS_FOREST_BLOOM_BITS; i++) {
 		if (atomic_read(&finf->oino_counted_bloom[i]))
 			set_bit_le(i, bb->bits);
 	}
-
 	scoutfs_block_put(sb, finf->oino_bl);
 	finf->oino_bl = NULL;
+
+	schedule_delayed_work(&finf->oino_work, msecs_to_jiffies(OINO_WORK_DELAY));
+
+	return 0;
+}
+
+void scoutfs_oino_delete_work(struct work_struct *work)
+{
+	struct forest_info *finf = container_of(work, struct forest_info,
+						oino_work.work);
+	struct super_block *sb = finf->sb;
+	struct scoutfs_net_roots roots;
+	SCOUTFS_BTREE_ITEM_REF(iref);
+	struct scoutfs_log_trees lt = {{{ 0 }}};
+	struct scoutfs_bloom_block *bb;
+	struct scoutfs_bloom_block *first_bb = NULL;
+	struct scoutfs_block *bl;
+	struct scoutfs_block *first_bl = NULL;
+	struct scoutfs_key ltk;
+	int ret = 0;
+	int i;
+
+	if (finf->shutdown)
+		return;
+
+	scoutfs_inc_counter(sb, forest_oino_delete_work);
+
+	ret = scoutfs_client_get_roots(sb, &roots);
+	if (ret)
+		goto out;
+
+	scoutfs_key_init_log_trees(&ltk, 0, 0);
+	for (;; scoutfs_key_inc(&ltk)) {
+		ret = scoutfs_btree_next(sb, &roots.logs_root, &ltk, &iref);
+		if (ret == 0) {
+			if (iref.val_len == sizeof(lt)) {
+				ltk = *iref.key;
+				memcpy(&lt, iref.val, sizeof(lt));
+			} else {
+				ret = -EIO;
+			}
+			scoutfs_btree_put_iref(&iref);
+		}
+		if (ret < 0) {
+			if (ret == -ENOENT)
+				break;
+			goto out; /* including stale */
+		}
+
+		if (lt.oino_bloom_ref.blkno == 0)
+			continue;
+
+		bl = read_bloom_ref(sb, &lt.oino_bloom_ref);
+		if (IS_ERR(bl)) {
+			ret = PTR_ERR(bl);
+			goto out;
+		}
+
+		/* If more than one, merge later bloom filters into the first */
+		if (!first_bl) {
+			first_bl = bl;
+			first_bb = first_bl->data;
+		} else {
+			bb = bl->data;
+
+			/* Use oldest block's seq */
+			first_bb->hdr.seq = cpu_to_le64(
+				min(le64_to_cpu(bb->hdr.seq),
+				    le64_to_cpu(first_bb->hdr.seq)));
+
+			for (i = 0; i < SCOUTFS_FOREST_BLOOM_ARRAY_LEN; i++) {
+				first_bb->bits[i] |= bb->bits[i];
+			}
+
+			scoutfs_block_put(sb, bl);
+		}
+	}
+
+	if (first_bb) {
+		ret = scoutfs_scan_orphans(sb, first_bb);
+		if (ret > 0) {
+			schedule_work(&finf->oino_work.work);
```

Comment:
Hmm, I don't think cancel_delayed_work_sync() can necessarily wait for normal scheduled work?  there's very confusing object layering going on in there.  I'd use schedule_delayed_work(, 0) just to be safe.